### DISCUSSION TITLE
Update from Kim

### DISCUSSION THREAD

**Discussion by websitecareio** [uldtot/wcioShop#12] [General]
This is an update post. I will here write recent updates instead of making a new post every time there is something.

**Comment by websitecareio** [comment]
My work today was mostly trying to make my own shop a bit better, getting texts and images ready, etc. buy I did work on some files too. 

- I fixed the issue with 404 when trying to visit anything other than the homepage. This was done with htaccess
- Fixed the naming for store catalog mode. I was using settingCatalogMode instead of settingStoreCatalogMode
- Began working on the admin panel. I need a way to add products and categories to continue.

Admin does not use pretty URLs as the front end does. Simply not needed and is a waste of time.

- Took a screenshot of the admin panel. Could not remember that I had made something that good looking :) YOu can view it here: https://github.com/websitecareio/wcioShop/blob/master/admin-demo.png

ave not yet committed the changes i made, no point at the moment. WIll collect a bit more changes before committing :)

**Comment by websitecareio** [comment]
Lots of work today. Just one thing committed (only to commit something today).
I was planning to make adding product work today, but just had to do a small thing with settings before that, ended up taking a couple of hours, and I haven't started on products yet. 

The changes I made were mostly to index and login validation really, needed a working structure that will work for every admin page. Hopefully, I have successfully thought of every scenario now and will not have to redo it again later. 

**Here are the order stuff loads in when admin pages are in use ( at the moment ):**

Lets assume we are on settings page, then wcio_settings.php are loaded and here is what is happening:
- Settings.tpl will be set as template file
- Index.php will be included for database, smarty functions and shop settings to be loaded.
- Then user is validated in index.php file. If error are found. Redirect to login.
- Then it will check if the URL is the dashboard and if it is, then it will include template functions.

Because it was settings we are looking aat this happens:
- We have a function to check if a setting group have been updated with new content, if there is, its save.
- After all other functions are done (saving, updating, logging, etc.)
- Now it loads tmplate functions to include functions needed in the template file (this includes getting the settings data)
- Sends template and data to smarty.

Why am i loading template functions in this way?
The reason is to make sure if we are updating data we wont be neeindg to fetch the data twice. So by checking the template file in index.php and including template function if its index.tpl and when its not index.tp inclde it after all other stuff happens. Kind of a middleware we made here.


![settings](https://github.com/websitecareio/wcioShop/assets/5160115/1980595a-f7d7-4e35-80ed-639c0d088625)

**Comment by websitecareio** [comment]
Small update before today's work. I was thinking I want to make the system useful in catalog mode, meaning I will focus on making the creation of products and categories work first, then pages. 
With these three parts ready, it would be possible to use the system in catalog mode and make a 1.0.0 release.

**Comment by websitecareio** [comment]
Hi! No update yesterday, had a very long day at work, and at 00:30 I decided to go to bed and stop at bit earlier today. 

Today's updates are a bit exiting because I have been working on committing a lot of stuff that isn't in the repository yet. 
My first task was to secure the database connection. I have done this by adding a .ini file outside of the accessible part of the server and then in the /inc/db.php file i made a change to have it look for the .ini file (i called the file private and you need to have the following content: 

[database_connection]
dbname = xxxxx
dbhost = xxxxx
dbuser = xxxxx
dbpass = xxxxx
dbport = xxxxx

and if the file is not present, you can use the variable instead. 
For me, this will make it easier to make commits because I do not have to check for this every time. Dont, want to accidentally share my db pass. 

And today i also started working on the part in admin to edit products, it's tied into creation and saving too, and some of it is the same code. 
Keeping it simple, can always add features later, I just need something that works at the moment. 

**Images**
I am wondering about images. Originally it was a mess, but I would love to have an image upload field with the option to drag the images around, and the first one there would then be a featured image. But I'm not sure how to do this, will see what things are available in javascript.

**Prices**
Another thing I am wondering about at the moment is prices. Let us assume I sell my products in DKK currency, but at a later point want to sell them in EUR. I could add some functions to automatically calculate this. But this way can cause issues for some and most would most likely want to set their own prices based on currency. An automatic option can easily be made with a JS function later.

My solution at the moment is when editing a product you will get the currencies you have selected in the settings shown (input field for each) and you can set the prices and at a later point, you get JS to calculate the rest if you want to. Then when saving the prices it will be saved into a separate table. I think this is the way to do this to avoid having to redo it later.

Example:
wcio_se_prices
productId,currency,value
xxx,DKK,99
xxx,SEK,125
xxx,EUR,15
xxx,USD,16
etc.

This is how the products part of admin looks like at the moment. Didn't get enough done today. Think I might need to redo some code later, not really sure how this should be done most efficiently...  
![products](https://github.com/websitecareio/wcioShop/assets/5160115/4025ad7c-dbd6-4c2e-9d10-c6b678f5a643)

![productedit](https://github.com/websitecareio/wcioShop/assets/5160115/e6a86bb6-b864-4388-91d1-4e5eb9b35a30)

**Comment by websitecareio** [comment]
Just a minor update today. It is Sunday after all :) 
In the coming week, I hope to be able to update and add new products. The images part I do not expect to have finished, still haven't found a way to upload images.

**Comment by uldtot** [comment]
So.... it has been a long time ... Today I worked on this again and have added some changes and new files. Can't guarantee anything but at the moment I have some time to work on it and some new motivation to do it. 

I have decided to make it even more simple for now, cutting down on functions (in order to actually make something that works)...

here is how the categories will be handled:
![image](https://github.com/user-attachments/assets/3458bd6c-556d-41d7-a476-ff35cffea646)

I will be working on categories first. When I can add, edit and remove categories and its descriptions I will continue with something. Think it will be pages, but not sure.

**Comment by uldtot** [comment]
I wanted to share some exciting updates and improvements we've made to the permalink functionality and SEO URL handling in our project. Here’s a breakdown of what’s new:

1. Enhanced URL Sanitization
We've introduced a new function, sanitizeSeoUrl(), which cleans up SEO URLs by:

Trimming whitespace: Removes any leading or trailing spaces.
Character Conversion: Converts special characters like æ, ø, and å into their ASCII equivalents (ae, oe, and aa).
Lowercasing: Ensures all URLs are in lowercase for consistency.
Allowed Characters: Only alphanumeric characters, dashes, underscores, and slashes are retained.
Hyphen Management: Spaces are replaced with hyphens, and duplicate hyphens are removed.
2. Improved Permalink Functionality
The savePermalink() function has been updated to:

Trim slashes: Removes leading and trailing slashes from the input URL, ensuring a clean format.
Suffix Logic: Suffixes (like -1, -2, etc.) are now added only if the same URL exists for a different post. This prevents unnecessary suffixes for the same post's URL.
URL Format: The new URL format now consistently includes slashes at the beginning and end.
3. URL Uniqueness Check
We’ve improved the logic for checking existing URLs in the wcio_se_permalinks table. The check now considers different post IDs, ensuring that conflicts are only counted for URLs associated with other posts.

Example Usage
You can now easily sanitize URLs and manage permalinks with these updates, making your SEO management smoother and more efficient.

Thanks for your support!

![image](https://github.com/user-attachments/assets/0161b03a-2bd8-4b03-b865-c192240003ae)
![image](https://github.com/user-attachments/assets/84898951-2d52-4eef-8bf1-08bdf6b251b1)

**Comment by uldtot** [comment]
Today I can now say something is ready! Categories can now be added, edited, and removed from the admin dashboard. So happy i made this work now.

Next, I will work on pages. It should not take too long, since it's mostly the same as categories. 

**What order am i working on this?**

- Categories 
I selected categories first because without them, products have no place to go.

-Pages
I could have chosen something else, but i wanted to make some content pages. Like the frontpage and "About us" setup on my demo site. and this felt like the fastest thing i could get ready, so i do not feel like I'm not getting anywhere. 

- Products
 I am thinking of products after pages. This will be a hard one, as there are many functions and lots of stuff to do, such as uploading images and field data. 
 But I think this is the right thing to work on because this will give me a store in catalog mode only if I get categories, pages, and products working. 

After the products, I have no idea what I am going to work on. There are no orders to work on, and no way to place an order. So I cannot be that. 
Then there is the cart function. Cart functions require a lot more than just a cart.. Shipping options, payment options, VAT calculation, forms to be filled out etc. So much stuff.  And i haven't figured out how to do the shipping and payment in a way other people can make their own payment options if they want to. 
So for now I do not know what comes after products.

**Comment by uldtot** [comment]
Another update.
I have made a file manager for this. The plan was to make it possible to upload images to products, but I ran into issues with how to store them in the DB, so decided to make a file upload first and combine the functions somehow later.

My thought right now is to use the file manager as a modal when selecting images and then when clicking the image in the file manager it will make a row in the DB with the URL etc. Same goes for gallery images and other file type. 

So I need to make this work for anything

I have decided to make it only possible to upload, rename and delete file and folders in the uploads directory. 

Todays work have not been added to the code yet, need some more changes before I add it. 

![image](https://github.com/user-attachments/assets/9d1b3e25-1bec-4dad-8f8f-bea50010a1eb)

**Comment by uldtot** [comment]
Another update. Loon time no see, I know it. 
It's hard to find the time to make stuff like this when I don't really need it myself. I am working on a solution to that by creating an actual store with real products I can see, so I need this again. 

The update here is that pages are working now. Didnt really do anything, just checked it again since its been soo long
<img width="903" height="832" alt="image" src="https://github.com/user-attachments/assets/96252c51-913c-45b3-9ee5-372fa6ad31eb" />

**Comment by uldtot** [comment]
An update here. I will be splitting the update posts into each own, so people can comment on a specific post instead of one with everything, it will get messy when people comment. Avoiding any issues here if this ever gets done and gets used :)
